File: apps/web/src/lib/slash-commands.ts
export interface SlashCommand {
  name: string;
  args?: string;
  description: string;
  category: 'session' | 'meta' | 'tools' | 'ops' | 'external';
}

export interface ParsedSlashCommand {
  raw: string;
  command: string;
  args: string[];
  options: Record<string, string | boolean>;
}

export const COMMANDS: SlashCommand[] = [
  { name: 'new', description: 'Start a new session', category: 'session' },
  { name: 'clear', description: 'Clear current session', category: 'session' },
  { name: 'compact', description: 'Trigger context compression', category: 'session' },
  { name: 'compress', description: 'Alias for /compact', category: 'session' },
  { name: 'undo', description: 'Remove last assistant turn', category: 'session' },
  { name: 'retry', description: 'Re-generate last assistant turn', category: 'session' },
  { name: 'help', description: 'List all commands', category: 'meta' },
  { name: 'version', description: 'Show version', category: 'meta' },
  { name: 'status', description: 'Show health summary', category: 'meta' },
  { name: 'debug', description: 'Toggle debug overlay', category: 'meta' },
  { name: 'tools', description: 'Open tools page', category: 'tools' },
  { name: 'skills', description: 'Open skills page', category: 'tools' },
  { name: 'model', args: '<model_id>', description: 'Switch model', category: 'tools' },
  { name: 'memory', description: 'Open memory page', category: 'tools' },
  { name: 'persona', description: 'Open persona editor', category: 'tools' },
  { name: 'usage', description: 'Show usage card', category: 'ops' },
  { name: 'fast', description: 'Switch to a faster model', category: 'ops' },
  { name: 'web', args: '<query>', description: 'Web search', category: 'external' },
  { name: 'image', args: '<prompt>', description: 'Image generation', category: 'external' },
  { name: 'browse', args: '<url>', description: 'Browse URL', category: 'external' },
  { name: 'code', args: '<task>', description: 'Code generation', category: 'external' },
  { name: 'shell', args: '<cmd>', description: 'Shell exec through allowlist', category: 'external' },
];

export function parseSlashCommand(text: string): ParsedSlashCommand | null {
  const raw = text.trim();
  if (!raw.startsWith('/')) return null;
  const parts = tokenize(raw.slice(1));
  if (parts.length === 0) return null;
  const command = parts[0] ?? '';
  const args: string[] = [];
  const options: Record<string, string | boolean> = {};
  for (let i = 1; i < parts.length; i += 1) {
    const token = parts[i] ?? '';
    if (token.startsWith('--') && token.length > 2) {
      const key = token.slice(2);
      if (key.includes('=')) {
        const [name, ...rest] = key.split('=');
        if (name) options[name] = rest.join('=');
      } else if (parts[i + 1] && !parts[i + 1]?.startsWith('--')) {
        options[key] = parts[i + 1] ?? '';
        i += 1;
      } else {
        options[key] = true;
      }
    } else {
      args.push(token);
    }
  }
  return { raw, command, args, options };
}

export function slashMatches(input: string): SlashCommand[] {
  if (!input.startsWith('/')) return [];
  const needle = input.slice(1).split(/\s+/, 1)[0]?.toLowerCase() ?? '';
  return COMMANDS.filter((cmd) => cmd.name.startsWith(needle)).slice(0, 8);
}

function tokenize(input: string): string[] {
  const out: string[] = [];
  let cur = '';
  let quote: '"' | "'" | null = null;
  for (let i = 0; i < input.length; i += 1) {
    const ch = input[i] ?? '';
    if (quote) {
      if (ch === quote) quote = null;
      else cur += ch;
    } else if (ch === '"' || ch === "'") {
      quote = ch;
    } else if (/\s/.test(ch)) {
      if (cur) {
        out.push(cur);
        cur = '';
      }
    } else {
      cur += ch;
    }
  }
  if (cur) out.push(cur);
  return out;
}

File: apps/server/api/slash_commands.py
"""Server-side slash command parser and stub executor — Phase 16."""

from __future__ import annotations

import shlex
from dataclasses import dataclass
from http import HTTPStatus

from . import __version__, auth as auth_module
from .config import Config
from .health import get_health
from .router import Request, Response, Router

COMMAND_NAMES = [
    "new", "clear", "compact", "compress", "undo", "retry",
    "help", "version", "status", "debug",
    "tools", "skills", "model", "memory", "persona",
    "usage", "fast", "web", "image", "browse", "code", "shell",
]


@dataclass(frozen=True)
class ParsedSlashCommand:
    raw: str
    command: str
    args: list[str]
    options: dict[str, str | bool]

    def to_dict(self) -> dict:
        return {"raw": self.raw, "command": self.command, "args": self.args, "options": self.options}


class SlashCommandError(ValueError):
    def __init__(self, code: str, detail: str | None = None) -> None:
        super().__init__(detail or code)
        self.code = code
        self.detail = detail or code


def parse_slash_command(text: str) -> ParsedSlashCommand:
    raw = text.strip()
    if not raw.startswith("/"):
        raise SlashCommandError("not_slash_command")
    try:
        parts = shlex.split(raw[1:])
    except ValueError as exc:
        raise SlashCommandError("invalid_slash_syntax", str(exc)) from exc
    if not parts:
        raise SlashCommandError("command_required")
    command, tokens = parts[0], parts[1:]
    if command not in COMMAND_NAMES:
        raise SlashCommandError("command_unknown", command)
    args: list[str] = []
    opts: dict[str, str | bool] = {}
    i = 0
    while i < len(tokens):
        tok = tokens[i]
        if tok.startswith("--") and len(tok) > 2:
            key = tok[2:]
            if "=" in key:
                k, v = key.split("=", 1)
                opts[k] = v
            elif i + 1 < len(tokens) and not tokens[i + 1].startswith("--"):
                opts[key] = tokens[i + 1]
                i += 1
            else:
                opts[key] = True
        else:
            args.append(tok)
        i += 1
    return ParsedSlashCommand(raw=raw, command=command, args=args, options=opts)


def execute_stub(parsed: ParsedSlashCommand) -> dict:
    if parsed.command in {"compact", "compress"}:
        return {"kind": "navigate", "to": "/rag", "text": "Open /rag to compact or search a saved session."}
    if parsed.command == "usage":
        return {"kind": "navigate", "to": "/usage", "text": "Open /usage for token and cost rollups."}
    if parsed.command == "version":
        return {"kind": "system", "text": __version__}
    if parsed.command == "help":
        return {"kind": "system", "text": "/" + ", /".join(COMMAND_NAMES)}
    if parsed.command == "model":
        if not parsed.args:
            return {"kind": "system", "text": "Usage: /model <model_id> [--temp 0.7]"}
        return {"kind": "system", "text": f"Model switched to {parsed.args[0]}", "model": parsed.args[0], "options": parsed.options}
    return {"kind": "intercept", "text": f"/{parsed.command} is registered; use the matching UI page or integration to execute it."}


def register_routes(cfg: Config) -> Router:
    router = Router()

    @router.route("GET", "/api/slash/commands")
    def _commands(req: Request) -> Response:
        if auth_module.authenticate(req, cfg) is None:
            return Response(HTTPStatus.UNAUTHORIZED, {"error": "not_authenticated"})
        return Response(HTTPStatus.OK, {"commands": COMMAND_NAMES})

    @router.route("POST", "/api/slash/parse")
    def _parse(req: Request) -> Response:
        if auth_module.authenticate(req, cfg) is None:
            return Response(HTTPStatus.UNAUTHORIZED, {"error": "not_authenticated"})
        try:
            body = req.json()
            parsed = parse_slash_command(str(body.get("text") or ""))
            return Response(HTTPStatus.OK, parsed.to_dict())
        except SlashCommandError as exc:
            return Response(HTTPStatus.BAD_REQUEST, {"error": exc.code, "detail": exc.detail})
        except ValueError:
            return Response(HTTPStatus.BAD_REQUEST, {"error": "invalid_json"})

    @router.route("POST", "/api/slash/execute")
    def _execute(req: Request) -> Response:
        if auth_module.authenticate(req, cfg) is None:
            return Response(HTTPStatus.UNAUTHORIZED, {"error": "not_authenticated"})
        try:
            parsed = parse_slash_command(str(req.json().get("text") or ""))
            if parsed.command == "status":
                return Response(HTTPStatus.OK, {"kind": "system", "health": get_health()})
            return Response(HTTPStatus.OK, execute_stub(parsed))
        except SlashCommandError as exc:
            return Response(HTTPStatus.BAD_REQUEST, {"error": exc.code, "detail": exc.detail})
        except ValueError:
            return Response(HTTPStatus.BAD_REQUEST, {"error": "invalid_json"})

    return router
